### ISSUE TITLE
Controls are mirroring "paper" process too much, doesn't adhere to standards

### ISSUE THREAD

**Issue by NoahKunin** [opencontrol/compliance-masonry#2]
After seeing many failed attempts to do what we're doing in the past, here's one thing we can intervene on now - the current commits too strongly imitate paper processes that have nothing to do with operational security, how other programs would ingest this data, and create deviations from the NIST 800-53 structure.

For example, a single control has sub-sections with arbitrary letters: a, b, c, d, that don't actually mean anything. That's just information overhead. Similarly, keying on `endnote` doesn't make any sense.

Here's the practical way to look at it - any control like `AC-X` and its subsequent enhancements `AC-X (Y)` per the NIST 800-53 format can either be a technical control, an organizational control, or some blend of the two.

If it's a technical control, there should be a space to include a hyperlink to *something* that governs that control - an example being the `UAA` module in Cloud Foundry.

If it's an organizational control, similarly there should *still* be a hyperlink to the policy or procedure governing that control. 

This also points at a potentially more flexible data structure. Instead of driving the hierarchy from NIST 800-53, we can treat each NIST control as a piece of metadata or tag to any particular technical or organizational control we have. As a result, we divorce ourselves from strictly linear inheritances that causes problem when a single technical or organizational "thing" actually helps us meet multiple controls. 

Further, this would allow us ingest the data in a way that better visualizes how the system actually functions, as a layers of interacting security boundaries and surfaces.

**Comment by dlapiduz** [comment]
To be honest, we started with the SSP and we structured stuff to mirror it as you said. That said it is easier to remove structure later than having to add it.

We are incorporating some flexibility in the way controls are compounded and I'd like to use that to create stuff that "makes sense" so I guess it is up to us on how to structure it.

Do you have any concrete ideas on how to make this more flexible?

**Comment by NoahKunin** [comment]
Shoot, @mogul and I had a meeting and looks like the tab died!

@dlapiduz yes. Will be in next PR.

**Comment by mogul** [comment]
Yep, @NoahKunin and I had a brainstorming session and I even scratched some things on a piece of paper... 
![img_20150909_170645 152 1](https://cloud.githubusercontent.com/assets/47762/9827981/b1c71662-589c-11e5-8e2e-d2f09cc9da7e.jpg)

So something like (sorry, I am new to YAML schema design): 
```yaml
control:
  name: Cloud Foundry User Authentication and Authorization (UAA)
  narrative: This is a separate component that blahblahblah. It segments access by blahblahblah.
  references: 
    - name: UAA design doc
      url: https://asdfasdf
    - name: Some other doc
     url: https://boobarbazbat
  governors:
    - name: UAA configuration
      url: https://pathtogitrepohead
    - name: Live test results
      url: https://dashboardwithupdatedtestresults
  satisfies:
    - standard: NIST-800-53:C-2
    - standard: PCI:Regulation6
```
where we have some other set of YAML which defines what all the standards are, eg in NIST-800-53.yml: 
```yaml
standard:
  - name: User Access
  - identifier: C-2
  - description: There is an affordance for managing access by...
```

**Comment by joshuamckenty** [comment]
I think it's a bit hazardous to head in this direction (component-level mappings rather than functional-level mappings). But I like declaring "satisfies" on the control itself, rather than including the function details (repetitively) under each control.

**Comment by joshuamckenty** [comment]
I've been looking for a formal meta-approach and I think the closest thing is probably "Literate Programming" by Knuth; in the sense that the same source needs to produce both executable binaries (or OpenSCAP compatible audit files, or CADF formatted mappings) and the "Book" that is the SSP. 

But "controls" in this context aren't functions, or components - they're reference architectures (including config, code, and operational guidelines.) So in that sense they feel more narrative than technical.

**Comment by mogul** [comment]
For terminology explanation, I was suggesting:
- “control” (thing in the system or process in the organization that is being described)
- “governor” (link to inspect how it’s actually configured to run, live... config files, live tests, kanban board, role roster, etc)
- “satisfies” (what standards it actually has to do with, which themselves are incorporated across a wide range of certifications, see for example FISMA Moderate and FEDRAMP and PCI) 
- “standard” for one particular element from one particular certification that has an externally-significant identifier (eg each of the ID's things in 800-53)

> But I like declaring "satisfies" on the control itself, rather than including the function details (repetitively) under each control.

Yeah, that's following Noah's lead to apply the DRY principle here. He suggested that we take the XML version of 800-53 and boil it down into YAML as input. Here's another scratching, and hey, JM in that picture is YOU, @joshuamckenty!

![img_20150909_173205 105](https://cloud.githubusercontent.com/assets/47762/9828387/7c6d48c4-58a1-11e5-853f-d6238ad67f36.jpg)

It's capturing that the standard-centric/GitBook-feeding narratives in the cloudfoundry-community/compliance repo would just be another output format. So rather than work directly there, we'd collaborate on an upstream cf.yaml which can be fed directly into control-masonry, with or without other controls specified for the org, the app, the local variant of CF, etc. Then the GitBook output sits parallel with a gov-specific SSP format, an interactive clickable summary for the cloud.gov landing page, etc. etc.

I get what you mean about incorporating SCAP, though... Because then one rendered output would be a testable config for auditing purposes. Nice!

**Comment by mogul** [comment]
Oh, and: 

> - “satisfies” (what standards it actually has to do with, which themselves are incorporated across a wide range of certifications, see for example FISMA Moderate and FEDRAMP and PCI)

We have a separate matrix being worked on that links individual controls in 800-53 to these various other standards. So as part of the output, we can generate a list of gaps toward assorted certifications... "You need something for NIST:800-53:BLAH in order to be ready for FEDRAMP, you need something in NIST:800-53:BAZ to be PCI compliant", etc.

**Comment by mogul** [comment]
One last thing: In the existing process, there's a Google spreadsheet being used to capture all this information for cloud.gov in matrix form... Another output format we thought of was CSV, from which we could regenerate that spreadsheet/matrix, which would make it easy to shift the source-of-truth toward YAML+git going forward.

**Comment by joshuamckenty** [comment]
Love it.

Regarding mapping 800-53 to different certifications, you should read up on http://cloudaudit.org/CloudAudit/Home.html. That was an earlier stab (led by Chris Hoff initially, although I sat on the working group for a while) at this type of mapping.

**Comment by mogul** [comment]
OK, so here's the cleaned up diagram:
![control-masonry](https://cloud.githubusercontent.com/assets/47762/9829499/08d2b1dc-58bb-11e5-8185-5dc617188ae7.png)

(It's not clear to me how I can attach a .gliffy file, but I have it if you need the source for this.)

Make sense?

**Comment by joshuamckenty** [comment]
I guess I don’t get why you would build a new tool for control-masonry instead of using GitBook, and writing some new Parsers and output formats.

**Comment by mogul** [comment]
Well, same as we discussed the other day... Not structured enough, not lint-able enough... If there's a way to do that with GitBook, please point me at it, as I'm totally unfamiliar with it!

**Comment by joshuamckenty** [comment]
I agree on the Gitbook spec, in terms of markdown and/or asciidoc. But GitBook is open source, and the parsers are reasonably isolated (https://github.com/GitbookIO/gitbook-parsers <https://github.com/GitbookIO/gitbook-parsers>). If we’re going to invent a yaml schema, we could still use this for processing and pipelining.

**Comment by mogul** [comment]
That would make sense if gitbook was flexible on output, but I'm not seeing many renderer plugins beyond things that embed eg rendered mathjax, web sequence diagrams, etc... The output is still a fundamentally a website or eBook format in those cases.

**Comment by joshuamckenty** [comment]
Well, it looks like this would be pretty easy to extend:

https://github.com/GitbookIO/gitbook/tree/master/lib/generators <https://github.com/GitbookIO/gitbook/tree/master/lib/generators>

**Comment by mogul** [comment]
Three of the output formats we've been able to anticipate are not fundamentally book-like... I'm not convinced that going through gitbook as an authoring/hub format would buy us much! I can see that we might want to make use of gitbook's plugins to similarly process embedded diagrams, etc, though.

OTOH there are definitely some benefits to using semi-structured Markdown as the authoring format, rather than YAML, since a lot of this will be narrative... We just need to agree on what schema that semi-structured content is trying to represent, and how it gets matched together to generate what we're after.

I'm also half-expecting @NoahKunin will whip a fully-formed pull-request out of his hat... :) Will wait to see if anyone else has any input Monday.

**Comment by dlapiduz** [comment]
I am a bit at a loss trying to follow the discussion here. Some thoughts so far:
- I agree that component level controls might not be a good idea... they are changed way too often and would make compliance brittle.
- This project and https://github.com/cloudfoundry-community/compliance are complimentary. As @mogul shows on the diagram, you would pull a shared control yml and incorporate it to get a full SSP.
- The discussion of gitbook vs other formats seems to be off-topic for this thread.

IMHO, I think that we need to start with something close to paper with a flexible enough format to switch it around when the time comes to be more streamlined. I think it would be pretty useful if we could get the cloud.gov SSP on a yml format all the way to the controls.

**Comment by NoahKunin** [comment]
@dlapiduz Concur that GitBook vs `?` is a separate issue. (literally, could be another issue in this GitHub repo or elsewhere. In short, end-state SSP formats will be as numerous as there are agencies. I'll have time/energy to be opinionated on that after we get the data structure to a MVP. For now, everything everyone is doing is already light years better then what I normally have to read. :+1: 

Comment on functional vs component coming right after this comment.

**Comment by ramirezg** [comment]
@mogul, the chart you created is a fantastic overview of the process. I followed your yaml examples to create possible formats for storing and combining the controls, standards, and certifications.


## My Cloud.gov/18F Controls/ Your Controls Documentation 
Without being combined with the standards and certifications yamls the control yamls can be use to generate readable documentation w/ gitbook, etc... 
```yaml
# cloud.yaml
controls:
  - name: Cloud Foundry User Authentication and Authorization (UAA)
    narrative: This is a separate component that blahblahblah. It segments access by blahblahblah.
    references: 
      - name: UAA design doc
        url: https://asdfasdf
      - name: Some other doc
        url: https://boobarbazbat
    governors:
      - name: UAA configuration
        url: https://pathtogitrepohead
      - name: Live test results
        url: https://dashboardwithupdatedtestresults
    satisfies:
      - standard: NIST-800-53:C-2
      - standard: PCI:Regulation6
```

## Standards Documentation
```yaml
# nist-800-53.yaml
standards:
  C-2:
    name: User Access
    description: There is an affordance for managing access by...

# PCI.yaml
standards:
  Regulation-6:
    name: User Access PCI
    description: There is an affordance for managing access by...
```

## Certifications 
Empty yaml for creating certification documentation. Serve as a template for combining controls and standards yamls. 

```yaml
# Fisma.yaml
standards:
  nist-800-53:
    C-2:
    C-3:
  PCI:
    6:
```

## Yaml output
Centralized yaml for a specific certification, can be used to render matrix.csv, gitbook.md, ssp.docx... This is were we will be able to see if any pieces are missing. 
```yaml
# FISMA_Project.yaml
standards:
  nist-800-53:
    C-2:
      name: User Access
      description: There is an affordance for managing access by...
      justifications:
        - name: Cloud Foundry User Authentication and Authorization (UAA)
          narrative: This is a separate component that blahblahblah. It segments access by blahblahblah.
          references: 
            - name: UAA design doc
              url: https://asdfasdf
            - name: Some other doc
              url: https://boobarbazbat
          governors:
            - name: UAA configuration
              url: https://pathtogitrepohead
            - name: Live test results
              url: https://dashboardwithupdatedtestresults
    C-3:
      a:
  PCI:
    Regulation-6:
      name: User Access PCI
      description: There is an affordance for managing access by...
      justifications:
        - name: Cloud Foundry User Authentication and Authorization (UAA)
          narrative: This is a separate component that blahblahblah. It segments access by blahblahblah.
          references: 
            - name: UAA design doc
              url: https://asdfasdf
            - name: Some other doc
              url: https://boobarbazbat
          governors:
            - name: UAA configuration
              url: https://pathtogitrepohead
            - name: Live test results
              url: https://dashboardwithupdatedtestresults
```

**Comment by mogul** [comment]
There you go. :) I'm guessing we can already derive a `lightweight-ato.yml` "certification" list from the materials that are about to be submitted on our behalf by GSA IT.

**Comment by NoahKunin** [comment]
@joshuamckenty @dlapiduz: See what you mean, but I think that the main issue around component vs functional vs architecture mapping right now is more a matter of diction than disagreement (statistically, a fairly common occurrence :chart_with_upwards_trend:). 

First, need to ensure we're all talking about the same thing. In terms of specificity, I would understand there to be a spectrum from component to architecture, with specificity _decreasing_ along the way.

A *component* might be a gem, a library, or even a line or lines of code.

A *function* might be something like `audit review, analysis, and reporting`, something that technically needs multiple components in order to function. 

A *reference architecture* is the thing that we are normally producing for SSP evidence right now. At the Government's current (rare) best, these architectures are enhanced by diagrams and flows generated programmatically through various tools (everything from a SCAP scan to something like [VisualOps](http://visualops.io/)), but more often than not, the only thing describing the architecture is a narrative claim accompanied by a low-resolution screenshot of a GUI or command line. :sob:  

I'm not dogmatic as to specifically where we end up on this spectrum, but we have to move toward more specificity. 

The situation we're solving for right now is that in 95% of agencies, developers will build a thing, get it to MVP or 1.0, and then sit in interviews with compliance managers for endless rounds of producing evidence on whether or not any particular control is satisfied, where that evidence isn't so much brittle as it is completely worthless to validate the system is secure or not.

In one world, the problem is the insane amount of time that takes, and we could solve that problem by making the narrative process information collection better, and that's a rather straightforward solution as described.

But that's not the actual problem I care about - I don't want to play *that* game better, I want [to end it entirely](https://www.youtube.com/watch?v=T0K14HBSWw0&t=0m31s) by creating data structures that enable us to create continuous monitoring platforms "auto-magically" by ingesting this data. That's still a long way off (our true "north") but I think the improvements already made in this issue thread absolutely get us on the right path forward.

**Comment by joshuamckenty** [comment]
I got your meaning, now - I was hearing “component” in the sense of a collection of functions - e.g., less specific than a function, rather than more-so. My concept could be rightly called a “system component”, and yours a “software component”?

Given that we’re both talking about the same goal, which is “executable documentation” for continuous monitoring, I think we need to start talking about OpenSCAP / whatever we think will execute these data structures (or some transform thereof).

**Comment by gregelin** [comment]
This is AWESOME. Like @NoahKunin, I've been dreaming of automating compliance and artifacts, structured data, DRY principles, and sharing control write ups...

The term 'control'  seems to get used to refer to both `AC-2` and `UAA`. That got me thinking...

Doing FISMA is a process of **continually mapping one set of controls onto another set of controls**. What makes this is hard is the many-to-many relationships. We map map Cloud Foundry's UAA module to NIST RMF AC-2; and then we want to figure out what NIST RMF controls maps to which PCI control (or which CCE maps to which NIST RMF).

Maybe **`nist-800-53:AC-2` can be thought of as the "abstract" control and `cloud-foundry:UAA` as the "instance."** To extend the programming metaphor, `AC-2` is the interface and the `UAA` the implementation of the interface)

The reason we struggle with the control implementation and the documentation production is because we are constantly fighting the control-to-control, many-to-many hairball.

![control-hairball](https://cloud.githubusercontent.com/assets/24979/10026835/de23906c-6131-11e5-9768-d98bb1be8068.jpg)

If we put all the abstract controls in the repo, and the implementation, then as described in @ramirezg comment https://github.com/18F/control-masonry/issues/2#issuecomment-140161940, we could get generate mappings.

Perhaps we need something like [Gruntjs](http://gruntjs.com) to lint and generate the mappings, and various other YAML documents and artifacts.

Would Control-masonry be the Gruntjs for Compliance? Or is Control-masonry a bit larger and the Gruntjs for Compliance is part of it?

I think @mogul's diagram expresses this at a high level. But it is possible we may find out interim output exists between before the final artifacts can be rendered.

**Comment by openprivacy** [comment]
Thinking on Greg's diagram, I'm wondering: for an "abstract" control like AC-2, there may be multiple instances, e.g., one for AWS console (cloud), one for SSH sessions to the server (OS) and one for the user application (college-choice). Is the thinking that there would be multiple complete copies of the control-masonry, one for each domain, in which case there would seem to be a lot of repeated work?

It seems more natural to have be multiple yaml files, e.g. cloud.yaml, os.yaml, cc.yaml, with each implementing AC-2 as needed for it's domain. This then begs the question: what happens when multiple instance "implementations" want to share code, such as (say) admin only access to the database server and nginx. Hard to keep DRY...

**Comment by mogul** [comment]
@openprivacy If you look at [my diagram above](https://github.com/18F/control-masonry/issues/2#issuecomment-139703330) you'll see that there it shows multiple .yaml files in "your controls", each expected to represent a specific domain. The expectation is that AC-2 will be referenced in a `satisfies:` field in as many controls in as many of those files as necessary, but get collated together in various forms of AC-2-centered output. 

I think of the separate .yaml files as "layers", like you'd see in [semi-transparent overlay pages in an anatomy textbook](https://www.etsy.com/listing/45409293/anatomy-celluloid-transparency-overlays). I think that it's OK to decorate specific parts of rendered output as having come from a specific layer.  

In cases where there is real repetition, I think that the source layer would help distinguish why something is listed twice. There will be cases where that's exactly what you want, and there will be other cases where you will see the repetition in the output and maybe go back to rearrange/add layers to make the structure clearer.

**Comment by gregelin** [comment]
The below is a pretty complex diagram showing how MITRE, the folks behind SCAP, think the various SCAP XML standards work together. It seemed worth including in this thread to have visibility on how SCAP currently works on production systems. You can see the orientation toward systems deployed in production vs systems under development.

![msm_measurement_and_architecture_diagram_handout](https://cloud.githubusercontent.com/assets/24979/10232898/545b6dda-6859-11e5-8cbd-f6f46ad2fe95.png)

[Mitre making security measurable diagram](http://makingsecuritymeasurable.mitre.org/docs/MSM_Measurement_and_Architecture_diagram_handout.pdf)

**Comment by mogul** [comment]
That makes my head hurt, and whoever made it owes the acronym jar about $50. But I think I get it, and it's an interesting model to keep in mind; thanks for pointing it out.

**Comment by gregelin** [comment]
@mogul - To be clear, I am not advocating that acronym world. Just wanted to point it out as a reference.

**Comment by chasballew** [comment]
FWIW, we see the CSA's CCM and CAIQ cross-mappings being used more widely in industry. A good place to start if you're interested in a crosswalk between different infosec schemes:

https://cloudsecurityalliance.org/download/cloud-controls-matrix-v3-0-1/

**Comment by mogul** [comment]
Wow, @chasballew, awesome!  Just the kind of thing I was hoping to find already exists in the diagram above... We want to be able to generate a list of gaps, eg "here's what you need to fill in to also be compliant with X, Y, and Z..." not just  FISMA. (@ramirezg you may want to compare our current inputs with this to see whether we can make it an upstream data source.)

**Comment by ramirezg** [comment]
@chasballew, These mapping are great, clean and nicely formatted.

**Comment by geramirez** [comment]
I'd like to change the component schema of control masonry to:
- allow for references to another or the same component's governors/artifacts (for example sometimes control justifications should be able to reference policies)
- contain information about the implementation and documentation status (this would help with gap analysis)

Thoughts on this format? 

```yaml 
component: Component Name # Name of the component
system: Component System Name # Name of the system the component is in
documentation_complete: false # Manual check if the documentation is complete
references: # References on the component
- name: General Reference  # Name of the reference
  url: Reference Website # Url of the reference
  type: URL # type of reference (will affect how it's rendered in the documentation)
governors:
  General_Governor_1:
    name: General Governor # Name of the governor/artifact
    url: Governor Website  # URL of the governor/artifact
    type: URL # type of reference (will affect how it's rendered in the documentation)
satisfies: # Standards that this component affects and justifications on how it satisfies the standards
  NIST-800-53:
    AC-2:
      narrative: Justification in narrative form # Justification text
      implementation_status: partial # Status of implementation 
      references: # References to artifacts/ governors that serve as proof
        - system: Component System Name  # System name of the reference (can link to other systems / components)
          component: Component Name # Component name of reference
          governor: General_Governor_1 # The specific artifact that the reference links to
```

**Comment by mogul** [comment]
I can see the need for the governor references across controls, particularly for the organizational layer controls which might refer to configuration of some other technical control in use. However, I can’t really comment as to the method by which those references are captured in the YAML; I don’t know if there’s any YAML stylistic norm for how to specify that kind of thing.  

Is it possible to refer to a governor without having to also refer to system+component? Eg if they’re not specified, system+component are assumed to be “local”? If so then I like that better, because the YAML can stay cleaner except when it really does need to have a link across boundaries.

I think the explicit flag for “documentation_complete” feels weird… Is that new with this topic? 

- Does the "documentation" refer to this YAML? 
  - Generally, I'd avoid including explicit schema about the state/version of a piece of documentation in the documentation itself. I would expect that this is more a factor of where the control YAML is, eg on a branch. 
  - However, that may be asking too much of authors who are not familiar with git and would be asked to juggle that without any explicit editor help.
- Does the "documentation" refer to other references? 
  - If so I would expect it to be explicit on each of the references or governors.

**Comment by geramirez** [comment]
> Is it possible to refer to a governor without having to also refer to system+component? Eg if they’re not specified, system+component are assumed to be “local”?

Yes, it's definitely possible to do this. The references structure above isn't inherent to yaml, it's just a system the script can work with. 

> I think the explicit flag for “documentation_complete” feels weird… Is that new with this topic?

This flag was added on another update, but can be removed. I added it because it helps maintain a list of components deemed ready by a human reviewer. Complete automation wasn't telling the entire story. In any case, this is a mostly superficial flag and can be removed. I played with the idea of adding more complexity, but I think we get to the point were writing these yamls without an editor is too tedious.
